### PR TITLE
Implement candidate registration and TopStakers seat manager.

### PR DESCRIPTION
**Pull request by piotrm50**
This PR adds support for top staker committee selection as well as registering candidates for next epoch committee selection.

Candidates are registered whenever they issue a single block before Epoch Nearing Threshold in the previous epoch.

### CODE REVIEW

File: pkg/protocol/commitment_verifier.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ type CommitmentVerifier struct {
 }
 
 func NewCommitmentVerifier(mainEngine *engine.Engine, lastCommonCommitmentBeforeFork *model.Commitment) *CommitmentVerifier {
-	committeeAtForkingPoint := mainEngine.SybilProtection.SeatManager().Committee(lastCommonCommitmentBeforeFork.Slot()).Accounts().IDs()
+	committeeAtForkingPoint := lo.Return1(mainEngine.SybilProtection.SeatManager().CommitteeInSlot(lastCommonCommitmentBeforeFork.Slot())).Accounts().IDs()
```
Comment (alexsporn) [inline pkg/protocol/commitment_verifier.go]: `CommitteeInSlot(...)` can return nil now, so this looks dangerous if you ignore the bool return variable. (This pattern is repeated everywhere where `Commitee(...)` was used before)

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+// SeatManager is a sybil protection module for the engine that manages the weights of actors according to their stake.
+type SeatManager struct {
+	apiProvider iotago.APIProvider
+	events      *seatmanager.Events
+
+	committeeStore   *epochstore.Store[*account.Accounts]
+	onlineCommittee  ds.Set[account.SeatIndex]
+	inactivityQueue  timed.PriorityQueue[account.SeatIndex]
+	lastActivities   *shrinkingmap.ShrinkingMap[account.SeatIndex, time.Time]
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go]: should we refactor the activity tracking into an own file? This is kind of duplicated code between poa and topstakers now

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	sort.Slice(candidatePools, func(i, j int) bool {
+		if candidatePools[i].pool.PoolStake != candidatePools[j].pool.PoolStake {
+			return candidatePools[i].pool.PoolStake > candidatePools[j].pool.PoolStake
+		}
+
+		if candidatePools[i].pool.ValidatorStake != candidatePools[j].pool.ValidatorStake {
+			return candidatePools[i].pool.ValidatorStake > candidatePools[j].pool.ValidatorStake
+		}
+
+		// TODO: add more tie-breaking
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go]: maybe prioritize candidates that haven't added an endepoch to its staking feature?

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+		// two candidates never have the same account ID because they come in a map
+		return bytes.Compare(candidatePools[i].accountID[:], candidatePools[j].accountID[:]) > 0
+	})
+
+	// Create new Accounts instance that only included validators selected to be part of the committee.
+	accounts := account.NewAccounts()
+	for _, candidatePool := range candidatePools[:s.optsSeatCount] {
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go]: you better check we have more candidates than seats here

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	defer s.activityMutex.RUnlock()
+
+	return s.onlineCommittee
+}
+
+func (s *SeatManager) SeatCount() int {
+	s.committeeMutex.RLock()
+	defer s.committeeMutex.RUnlock()
+
+	return int(s.optsSeatCount)
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go:154]: same as before, what happens if we have less candidates?

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	return nil
+}
+
+func (s *SeatManager) SetCommittee(epoch iotago.EpochIndex, validators *account.Accounts) error {
+	s.committeeMutex.Lock()
+	defer s.committeeMutex.Unlock()
+
+	if validators.Size() != int(s.optsSeatCount) {
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go:194]: Again here, what would happen if we don't have enough candidates?

File: pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go [hunk 7]
```diff
@@ -108,8 +98,42 @@ func (o *SybilProtection) Shutdown() {
 	o.TriggerStopped()
 }
 
-func (o *SybilProtection) TrackValidationBlock(block *blocks.Block) {
-	o.performanceTracker.TrackValidationBlock(block)
+func (o *SybilProtection) TrackBlock(block *blocks.Block) {
+	accountData, exists, err := o.ledger.Account(block.ProtocolBlock().IssuerID, block.SlotCommitmentID().Slot())
```
Comment (alexsporn) [inline pkg/protocol/sybilprotection/sybilprotectionv1/sybilprotection.go:109]: why fetch the accountData before even checking if it is a candidacy announcement? Check first if we need to do the work or not.

File: pkg/storage/prunable/prunable.go [hunk 8]
```diff
@@ -167,7 +171,7 @@ func (p *Prunable) Rollback(targetSlot iotago.SlotIndex) error {
 			}
 
 			if shouldRollback {
-				if err := p.committee.DeleteEpoch(epoch); err != nil {
+				if err = p.committee.DeleteEpoch(epoch); err != nil {
```
Comment (alexsporn) [inline pkg/storage/prunable/prunable.go]: why this change? it is a local var captured only for the if.

File: pkg/storage/prunable/prunable_slot.go [hunk 9]
```diff
@@ -52,6 +53,11 @@ func (p *Prunable) RootBlocks(slot iotago.SlotIndex) (*slotstore.Store[iotago.Bl
 	), nil
 }
 
+func (p *Prunable) CommitteeCandidates(epoch iotago.EpochIndex) (kvstore.KVStore, error) {
+	// Slot 0 does not exist, but we need to add it to the prefix to avoid random clashes with other keys.
+	return p.prunableSlotStore.Get(epoch, byteutils.ConcatBytes(iotago.SlotIndex(0).MustBytes(), kvstore.Realm{epochPrefixCommitteeCandidates}))
```
Comment (alexsporn) [inline pkg/storage/prunable/prunable_slot.go]: this is hacky, better use the epoch startSlot so it at least matches the other stores.

**Review by alexsporn — CHANGES_REQUESTED** [review]
This kind of introduces a lot of new TODOs

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go (hunk 3 above)
Comment (piotrm50) [reply pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go]: I added some more things there. It wasn't there initially because the data wasn't available. Now I changed what RotateCommittee receives which imo makes it clearer.

File: pkg/protocol/commitment_verifier.go (hunk 1 above)
Comment (piotrm50) [reply pkg/protocol/commitment_verifier.go]: We need to discuss what to do if the committee does not exists for some reason, because this should not happen, but if it does then it's pretty severe.

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go (hunk 5 above)
Comment (piotrm50) [reply pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go:154]: This should always return the same number as this is the number of validators we will have at all times.

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go (hunk 6 above)
Comment (piotrm50) [reply pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go:194]: This method is only used for re-using the committee for the previous slot, so here we require the exact number of validators as we use them as they come.

File: pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go (hunk 4 above)
Comment (piotrm50) [reply pkg/protocol/sybilprotection/seatmanager/topstakers/topstakers.go]: Good catch! Changed it to re-use the committee from  the previous slot in this case as discussed before.

**Review by alexsporn — APPROVED** [review]

**Review by alexsporn — APPROVED** [review]
